File: tools/covenant/prepare_spawn.py
#!/usr/bin/env python3
"""Prepare Covenant spawn payload + prompt using identity v1 defaults.

Default path (enforced):
1) Normalize payload (optional legacy shim)
2) Validate handshake against identity registry
3) Build injected prompt from identity contract

Usage:
  python3 tools/covenant/prepare_spawn.py <handshake-or-legacy.json> [--output-dir <dir>] [--legacy-shim]
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any

WORKSPACE_ROOT = Path("/Users/hd/openclaw")
HANDSHAKE_VALIDATOR = WORKSPACE_ROOT / "tools" / "covenant" / "validate_spawn_handshake.py"
PROMPT_BUILDER = WORKSPACE_ROOT / "tools" / "covenant" / "build_identity_spawn_prompt.py"
ROUTER = WORKSPACE_ROOT / "tools" / "covenant" / "route_workflow.py"

DEFAULT_OUTPUT_FORMAT = {
    "type": "markdown",
    "sections": ["summary", "changes", "validation", "risks", "next_steps"],
}

DEFAULT_TIMEOUT_RETRY = {
    "timeout_seconds": 1800,
    "max_retries": 2,
    "retry_on": ["transient_tool_failure", "network_timeout"],
    "escalate_on": ["auth_failure", "permission_denied", "requirements_ambiguous"],
}

DEFAULT_CALLBACK = {
    "update_channel": "subagent_result_push",
    "final_channel": "requester_session",
    "heartbeat_interval_seconds": 300,
    "on_blocked": "immediate",
}

DEFAULT_CONSTRAINTS = {
    "workspace_root": str(WORKSPACE_ROOT),
    "allowed_paths": [str(WORKSPACE_ROOT)],
    "forbidden_actions": ["force_push", "destructive_delete", "external_message_without_approval"],
}


class PrepError(Exception):
    pass


def _load_json(path: Path, label: str) -> Any:
    if not path.exists():
        raise PrepError(f"{label} not found: {path}")
    try:
        return json.loads(path.read_text())
    except json.JSONDecodeError as exc:
        raise PrepError(f"{label} invalid JSON: {exc}") from exc


def _to_string_list(value: Any) -> list[str] | None:
    if isinstance(value, list):
        out = [str(v).strip() for v in value if str(v).strip()]
        return out or None
    if isinstance(value, str) and value.strip():
        return [value.strip()]
    return None


def normalize_payload(raw: dict[str, Any], legacy_shim: bool) -> tuple[dict[str, Any], list[str]]:
    if not legacy_shim:
        return raw, []

    normalized = dict(raw)
    notes: list[str] = []

    mission = normalized.pop("mission", None)
    expected_outcomes = normalized.pop("expected_outcomes", None)
    expected_outcome = normalized.pop("expected_outcome", None)

    if "objective" not in normalized and isinstance(mission, str) and mission.strip():
        normalized["objective"] = mission.strip()
        notes.append("mapped legacy field 'mission' -> 'objective'")

    if "success_criteria" not in normalized:
        criteria = _to_string_list(expected_outcomes or expected_outcome)
        if criteria:
            normalized["success_criteria"] = criteria
            notes.append("mapped legacy field 'expected_outcome(s)' -> 'success_criteria'")

    if "output_format" not in normalized:
        normalized["output_format"] = dict(DEFAULT_OUTPUT_FORMAT)
        notes.append("injected default 'output_format'")

    if "timeout_retry_policy" not in normalized:
        normalized["timeout_retry_policy"] = dict(DEFAULT_TIMEOUT_RETRY)
        notes.append("injected default 'timeout_retry_policy'")

    if "callback" not in normalized:
        normalized["callback"] = dict(DEFAULT_CALLBACK)
        notes.append("injected default 'callback'")
    elif isinstance(normalized.get("callback"), dict) and "update_channel" not in normalized["callback"]:
        normalized["callback"] = {**DEFAULT_CALLBACK, **normalized["callback"]}
        notes.append("filled missing callback.update_channel via compatibility defaults")

    if "constraints" not in normalized:
        normalized["constraints"] = dict(DEFAULT_CONSTRAINTS)
        notes.append("injected default 'constraints'")

    return normalized, notes


def _run(cmd: list[str]) -> None:
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        msg = (result.stderr or result.stdout).strip()
        raise PrepError(msg or f"command failed: {' '.join(cmd)}")


def maybe_auto_route_identity(payload: dict[str, Any], auto_route: bool) -> tuple[dict[str, Any], list[str]]:
    if not auto_route or payload.get("agent_identity_id"):
        return payload, []

    if not ROUTER.exists():
        raise PrepError(f"routing tool not found: {ROUTER}")

    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as tmp:
        tmp.write(json.dumps(payload))
        tmp_path = Path(tmp.name)

    try:
        result = subprocess.run(
            ["python3", str(ROUTER), "--plan", str(tmp_path)],
            capture_output=True,
            text=True,
        )
    finally:
        tmp_path.unlink(missing_ok=True)

    if result.returncode != 0:
        msg = (result.stderr or result.stdout).strip()
        raise PrepError(f"auto-route failed: {msg}")

    line = None
    for raw in (result.stdout or "").splitlines():
        if raw.startswith("ROUTING_PLAN_JSON:"):
            line = raw[len("ROUTING_PLAN_JSON:") :].strip()
            break
    if not line:
        raise PrepError("auto-route failed: missing ROUTING_PLAN_JSON output")

    try:
        route = json.loads(line)
    except json.JSONDecodeError as exc:
        raise PrepError(f"auto-route failed: invalid routing JSON: {exc}") from exc

    identity = route.get("primary_agent_identity_id")
    if not isinstance(identity, str) or not identity.strip():
        raise PrepError("auto-route failed: missing primary_agent_identity_id")

    updated = dict(payload)
    updated["agent_identity_id"] = identity
    notes = [f"auto-routed missing 'agent_identity_id' -> '{identity}'"]
    return updated, notes


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Prepare Covenant identity-v1 spawn artifacts")
    p.add_argument("payload", help="Path to handshake payload (or legacy payload with --legacy-shim)")
    p.add_argument("--output-dir", default="/tmp/covenant-spawn", help="Directory to write normalized handshake + prompt")
    p.add_argument("--legacy-shim", action="store_true", help="Apply compatibility shim for legacy payload shapes")
    p.add_argument("--auto-route", action="store_true", help="Infer agent_identity_id via route_workflow.py when missing")
    return p.parse_args()


def main() -> None:
    args = parse_args()

    try:
        payload_path = Path(args.payload).expanduser().resolve()
        output_dir = Path(args.output_dir).expanduser().resolve()

        raw = _load_json(payload_path, "payload")
        if not isinstance(raw, dict):
            raise PrepError("payload root must be an object")

        normalized, notes = normalize_payload(raw, args.legacy_shim)
        normalized, route_notes = maybe_auto_route_identity(normalized, args.auto_route)
        notes.extend(route_notes)

        output_dir.mkdir(parents=True, exist_ok=True)
        normalized_path = output_dir / "handshake.normalized.json"
        prompt_path = output_dir / "spawn.prompt.txt"

        normalized_path.write_text(json.dumps(normalized, indent=2) + "\n")

        _run(["python3", str(HANDSHAKE_VALIDATOR), str(normalized_path)])
        _run(["python3", str(PROMPT_BUILDER), str(normalized_path), "--output", str(prompt_path)])

        print(f"SPAWN_PREPARED: {output_dir}")
        print(f"HANDSHAKE_PATH: {normalized_path}")
        print(f"PROMPT_PATH: {prompt_path}")
        if notes:
            print("COMPAT_SHIM_APPLIED:")
            for n in notes:
                print(f"- {n}")
    except PrepError as exc:
        print(f"SPAWN_PREP_INVALID: {exc}", file=sys.stderr)
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: tools/covenant/prepare_spawn.ts
#!/usr/bin/env npx tsx

/** Prepare Covenant spawn payload + prompt using identity v1 defaults. */

import fs from "fs";
import os from "os";
import path from "path";
import { spawnSync } from "child_process";
import { resolveRepoPath } from "../lib/paths.js";

const WORKSPACE_ROOT = resolveRepoPath();
const HANDSHAKE_VALIDATOR = path.join(WORKSPACE_ROOT, "tools", "covenant", "validate_spawn_handshake.ts");
const PROMPT_BUILDER = path.join(WORKSPACE_ROOT, "tools", "covenant", "build_identity_spawn_prompt.ts");
const ROUTER = path.join(WORKSPACE_ROOT, "tools", "covenant", "route_workflow.ts");

const DEFAULT_OUTPUT_FORMAT = {
  type: "markdown",
  sections: ["summary", "changes", "validation", "risks", "next_steps"],
};

const DEFAULT_TIMEOUT_RETRY = {
  timeout_seconds: 1800,
  max_retries: 2,
  retry_on: ["transient_tool_failure", "network_timeout"],
  escalate_on: ["auth_failure", "permission_denied", "requirements_ambiguous"],
};

const DEFAULT_CALLBACK = {
  update_channel: "subagent_result_push",
  final_channel: "requester_session",
  heartbeat_interval_seconds: 300,
  on_blocked: "immediate",
};

const DEFAULT_CONSTRAINTS = {
  workspace_root: WORKSPACE_ROOT,
  allowed_paths: [WORKSPACE_ROOT],
  forbidden_actions: ["force_push", "destructive_delete", "external_message_without_approval"],
};

class PrepError extends Error {}

type Json = Record<string, any>;

function loadJson(filePath: string, label: string): any {
  if (!fs.existsSync(filePath)) {
    throw new PrepError(`${label} not found: ${filePath}`);
  }
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch (err) {
    throw new PrepError(`${label} invalid JSON: ${err instanceof Error ? err.message : String(err)}`);
  }
}

function toStringList(value: any): string[] | null {
  if (Array.isArray(value)) {
    const out = value.map((v) => String(v).trim()).filter((v) => v);
    return out.length ? out : null;
  }
  if (typeof value === "string" && value.trim()) return [value.trim()];
  return null;
}

function normalizePayload(raw: Json, legacyShim: boolean): [Json, string[]] {
  if (!legacyShim) return [raw, []];

  const normalized: Json = { ...raw };
  const notes: string[] = [];

  const mission = normalized.mission;
  const expectedOutcomes = normalized.expected_outcomes;
  const expectedOutcome = normalized.expected_outcome;

  delete normalized.mission;
  delete normalized.expected_outcomes;
  delete normalized.expected_outcome;

  if (!("objective" in normalized) && typeof mission === "string" && mission.trim()) {
    normalized.objective = mission.trim();
    notes.push("mapped legacy field 'mission' -> 'objective'");
  }

  if (!("success_criteria" in normalized)) {
    const criteria = toStringList(expectedOutcomes ?? expectedOutcome);
    if (criteria) {
      normalized.success_criteria = criteria;
      notes.push("mapped legacy field 'expected_outcome(s)' -> 'success_criteria'");
    }
  }

  if (!("output_format" in normalized)) {
    normalized.output_format = { ...DEFAULT_OUTPUT_FORMAT };
    notes.push("injected default 'output_format'");
  }

  if (!("timeout_retry_policy" in normalized)) {
    normalized.timeout_retry_policy = { ...DEFAULT_TIMEOUT_RETRY };
    notes.push("injected default 'timeout_retry_policy'");
  }

  if (!("callback" in normalized)) {
    normalized.callback = { ...DEFAULT_CALLBACK };
    notes.push("injected default 'callback'");
  } else if (typeof normalized.callback === "object" && normalized.callback && !("update_channel" in normalized.callback)) {
    normalized.callback = { ...DEFAULT_CALLBACK, ...normalized.callback };
    notes.push("filled missing callback.update_channel via compatibility defaults");
  }

  if (!("constraints" in normalized)) {
    normalized.constraints = { ...DEFAULT_CONSTRAINTS };
    notes.push("injected default 'constraints'");
  }

  return [normalized, notes];
}

function runCmd(cmd: string[]): void {
  const result = spawnSync(cmd[0], cmd.slice(1), { encoding: "utf8" });
  if (result.status !== 0) {
    const msg = (result.stderr || result.stdout || "").toString().trim();
    throw new PrepError(msg || `command failed: ${cmd.join(" ")}`);
  }
}

function maybeAutoRouteIdentity(payload: Json, autoRoute: boolean): [Json, string[]] {
  if (!autoRoute || payload.agent_identity_id) return [payload, []];

  if (!fs.existsSync(ROUTER)) {
    throw new PrepError(`routing tool not found: ${ROUTER}`);
  }

  const tmpPath = path.join(os.tmpdir(), `covenant-route-${process.pid}-${Date.now()}.json`);
  fs.writeFileSync(tmpPath, JSON.stringify(payload));

  let result;
  try {
    result = spawnSync(ROUTER, ["--plan", tmpPath], { encoding: "utf8" });
  } finally {
    try {
      fs.unlinkSync(tmpPath);
    } catch {
      // ignore
    }
  }

  if (!result || result.status !== 0) {
    const msg = ((result?.stderr || result?.stdout) ?? "").toString().trim();
    throw new PrepError(`auto-route failed: ${msg}`);
  }

  let line: string | null = null;
  for (const raw of (result.stdout || "").toString().split(/\r?\n/)) {
    if (raw.startsWith("ROUTING_PLAN_JSON:")) {
      line = raw.slice("ROUTING_PLAN_JSON:".length).trim();
      break;
    }
  }
  if (!line) throw new PrepError("auto-route failed: missing ROUTING_PLAN_JSON output");

  let route: Json;
  try {
    route = JSON.parse(line);
  } catch (err) {
    throw new PrepError(`auto-route failed: invalid routing JSON: ${err instanceof Error ? err.message : String(err)}`);
  }

  const identity = route.primary_agent_identity_id;
  if (typeof identity !== "string" || !identity.trim()) {
    throw new PrepError("auto-route failed: missing primary_agent_identity_id");
  }

  const updated = { ...payload, agent_identity_id: identity };
  return [updated, [`auto-routed missing 'agent_identity_id' -> '${identity}'`]];
}

function usageError(): never {
  console.error("usage: prepare_spawn.py <handshake-or-legacy.json> [--output-dir <dir>] [--legacy-shim] [--auto-route]");
  process.exit(2);
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const payloadArg = args[0];
  if (!payloadArg) usageError();

  const outputDir = (() => {
    const idx = args.indexOf("--output-dir");
    if (idx >= 0 && args[idx + 1]) return args[idx + 1];
    const eq = args.find((a) => a.startsWith("--output-dir="));
    if (eq) return eq.slice("--output-dir=".length);
    return "/tmp/covenant-spawn";
  })();

  const legacyShim = args.includes("--legacy-shim");
  const autoRoute = args.includes("--auto-route");

  try {
    const payloadPath = path.resolve(payloadArg);
    const outputPath = path.resolve(outputDir);

    const raw = loadJson(payloadPath, "payload");
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
      throw new PrepError("payload root must be an object");
    }

    let normalized: Json;
    let notes: string[];
    [normalized, notes] = normalizePayload(raw, legacyShim);
    let routeNotes: string[];
    [normalized, routeNotes] = maybeAutoRouteIdentity(normalized, autoRoute);
    notes = notes.concat(routeNotes);

    fs.mkdirSync(outputPath, { recursive: true });
    const normalizedPath = path.join(outputPath, "handshake.normalized.json");
    const promptPath = path.join(outputPath, "spawn.prompt.txt");

    fs.writeFileSync(normalizedPath, JSON.stringify(normalized, null, 2) + "\n");

    runCmd([HANDSHAKE_VALIDATOR, normalizedPath]);
    runCmd([PROMPT_BUILDER, normalizedPath, "--output", promptPath]);

    console.log(`SPAWN_PREPARED: ${outputPath}`);
    console.log(`HANDSHAKE_PATH: ${normalizedPath}`);
    console.log(`PROMPT_PATH: ${promptPath}`);
    if (notes.length) {
      console.log("COMPAT_SHIM_APPLIED:");
      notes.forEach((n) => console.log(`- ${n}`));
    }
  } catch (err) {
    if (err instanceof PrepError) {
      console.error(`SPAWN_PREP_INVALID: ${err.message}`);
      process.exit(1);
    }
    throw err;
  }
}

main().catch((err) => {
  console.error(err instanceof Error ? err.stack || err.message : String(err));
  process.exit(1);
});
